### PR TITLE
Dev Home SDK changes for FileExplorerSourceControlIntegration

### PR DESCRIPTION
**Pull request by ssparach**
## Summary of the pull request
Additions to the Dev Home SDK (as designed by @DefaultRyan) for the File Explorer Source Control Integration feature

## References and relevant issues
#2314 

## Detailed description of the pull request / Additional comments
Three new interfaces added: 
- ILocalRepository: This interface is responsible for obtaining and relaying information on the repository from source control technology. 
- GetLocalRepositoryResult: This interface defines the result objects that can be returned to Dev Home when it tries to obtain an enhanced repository (i.e. a repository for which this feature is enabled) 
- ILocalRepositoryProvider: This interface is responsible for obtaining the ILocalRepository object given a root path

## Validation steps performed
Local Build

## PR checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: extensionsdk/Microsoft.Windows.DevHome.SDK/GetLocalRepositoryResult.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace winrt::Microsoft::Windows::DevHome::SDK::implementation
+{
+    struct GetLocalRepositoryResult : GetLocalRepositoryResultT<GetLocalRepositoryResult>
+    {
+        GetLocalRepositoryResult() = default;
+
+        explicit GetLocalRepositoryResult(winrt::Microsoft::Windows::DevHome::SDK::ILocalRepository const& repository);
+        GetLocalRepositoryResult(winrt::hresult const& e, hstring const& displayMessage, hstring const& diagnosticText);
+        winrt::Microsoft::Windows::DevHome::SDK::ILocalRepository Repository();
```
Comment (DefaultRyan) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/GetLocalRepositoryResult.h:12]: Consider making these two prop getters `const`

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl [hunk 2]
```diff
@@ -1380,7 +1380,48 @@ namespace Microsoft.Windows.DevHome.SDK
...
 
+    // This interface is responsible for obtaining and relaying information on the repository
+    // from source control technology
+    [contract(Microsoft.Windows.DevHome.SDK.DevHomeContract, 5)]
+    interface ILocalRepository
+    {
+        // This function returns the requested information on a relative file path (under a
+        // tracked repository root folder) from the source control management technology.
+        // The requested information includes a predefined set of properties FE is aware of.
+        // These properties are Status, LastChangeDate, LastChangeAuthorEmail, LastChangeAuthorName,
```
Comment (DefaultRyan) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:1393]: Use the full names of these properties. Should also document what the expected types of these properties are.

Comment (DefaultRyan) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:1393]: Use the full names of these properties. Should also document what the expected types of these properties are. Consider adding documentation to `devhome/doc/extensions/readme.md`. Better yet, consider adding dedicated docs to a new folder e.g. `devhome/doc/extensions/LocalRepository`.

**Review by DefaultRyan — CHANGES_REQUESTED** [review]

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl [hunk 3]
```diff
@@ -1380,7 +1380,50 @@ namespace Microsoft.Windows.DevHome.SDK
...
 
+    // This interface is responsible for obtaining and relaying information on the repository
+    // from source control technology
+    [contract(Microsoft.Windows.DevHome.SDK.DevHomeContract, 5)]
+    interface ILocalRepository
+    {
+        // This function returns the requested information on a relative file path (under a
+        // tracked repository root folder) from the source control management technology.
+        // The requested information includes a predefined set of properties FE is aware of.
+        // These properties are System.VersionControl.Status, System.VersionControl.LastChangeDate,
```
Comment (dhoehna) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:1393]: This list is duplicated and has to stay in sync with the readme.

Maybe remove these properties and reference the readme, or vice-versa?

File: extensionsdk/Microsoft.Windows.DevHome.SDK/GetLocalRepositoryResult.cpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#include "GetLocalRepositoryResult.g.cpp"
+
+namespace winrt::Microsoft::Windows::DevHome::SDK::implementation
+{
+    GetLocalRepositoryResult::GetLocalRepositoryResult(winrt::Microsoft::Windows::DevHome::SDK::ILocalRepository const& repository) :
+        _repository(repository), _result(ProviderOperationStatus::Success, S_OK, hstring(), hstring())
+    {
+    }
+
+    GetLocalRepositoryResult::GetLocalRepositoryResult(winrt::hresult const& e, hstring const& displayMessage, hstring const& diagnosticText) :
```
Comment (dhoehna) [inline extensionsdk/Microsoft.Windows.DevHome.SDK/GetLocalRepositoryResult.cpp:12]: I believe DevHome will crash if Repository() is called after this constructor is invoked.

Comment (DefaultRyan) [reply extensionsdk/Microsoft.Windows.DevHome.SDK/GetLocalRepositoryResult.cpp:12]: `Repository()` can be called safely, but the result will contain a `nullptr`. Accessing members of the returned `ILocalRepository` without checking the return value *will* crash, as expected.

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl (hunk 3 above)
Comment (ssparach) [reply extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl]: Resolved in latest. Thanks!

File: extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl [hunk 5]
```diff
@@ -1380,7 +1380,50 @@ namespace Microsoft.Windows.DevHome.SDK
...
+        // The requested information includes a predefined set of properties FE is aware of.
+        // These properties are System.VersionControl.Status, System.VersionControl.LastChangeDate, 
+        // System.VersionControl.LastChangeAuthorEmail, System.VersionControl.LastChangeAuthorName,
+        // System.VersionControl.LastChangeID, System.VersionControl.LastChangeMessage, 
+        // System.VersionControl.CurrentFolderStatus
+        Windows.Foundation.Collections.IPropertySet GetProperties(String[] properties, String relativePath);
+    };
+
+    // This interface defines the result objects that can be returned to DevHome when it tries to
+    // obtain an enhanced repository
```
[the remark this answers is not in the document]
Comment (ssparach) [reply extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:1398]: This is the term given to repositories for which File Explorer can display property information from version control technologies.

[the remark this answers is not in the document]
Comment (dhoehna) [reply extensionsdk/Microsoft.Windows.DevHome.SDK/Microsoft.Windows.DevHome.SDK.idl:1398]: Can that description of an enhanced repository be added to the comments?

**Review by dhoehna — APPROVED** [review]

**Review by DefaultRyan — APPROVED** [review]

**Pull request merged**